Add documentation

aurelien-boch opened the pull request.

aurelien-boch committed:
docs: add documentation

Created empty markdowns in docs/interfaces.
Created readme.md in docs/

aurelien-boch committed:
feat: Add IClient and IListener doc

aurelien-boch committed:
feat: Add IModule documentation

Add IModule documentation, fix IListener and IClient errors

aurelien-boch committed:
feat: Fix interfaces description, so all looks the same
feat: Fix docs readme.md and add installation procedure

aurelien-boch committed:
doc: improve root readme.md
doc: improve documentation readme.md

aurelien-boch committed:
doc: improve documentation readme.md

rateau-violent commented on doc/interfaces/IClient.md:
```diff
@@ -0,0 +1,12 @@
+The goal of this interface is to abstract the communication method to the rest of the pipeline.
+
+### Listener
+This interface will be used in the Listener. You are free to chose how your listener work and how it will interact
```
how your listener work -> work**s**

rateau-violent commented on doc/interfaces/IListener.md:
```diff
@@ -0,0 +1,20 @@
+
+### Description
+The IListener interface defines methods that `Listener` must implement.
+The goal of the `listener` is to listen for new connections and wait for full packet before calling the callback method,
```
"listener" or "Listener", you have to choose

rateau-violent commented on doc/interfaces/IListener.md:
```diff
@@ -0,0 +1,20 @@
+The goal of the `listener` is to listen for new connections and wait for full packet before calling the callback method,
+that was provided using the `onPacket()` method.
+
+You are free to chose how will the internal listener work (as long as it matches your client implementation obviously).
```
how will the internal listener work -> how the internal listener will work

rateau-violent commented on doc/interfaces/IListener.md:
```diff
@@ -0,0 +1,20 @@
+and then send the decrypted version to the callback.
+
+### Client lifetime
+The listener is responsible for the lifetime of the `client`. It MUST it close it when necessary (for example:
```
**It** MUST **it** close **it** -> It must close it ?

aurelien-boch committed:
docs: updated readme and UML with new types

aurelien-boch committed:
docs: add const to the PipelineConfig in the configurePipeline method (architecture.puml)

rateau-violent commented on include/IPipelineManager.hpp:
```diff
@@ -0,0 +1,45 @@
+             */
+            void stop() = 0;
+
+        protected:
+            /**
+             * @brief Handles a client request by calling the parser, the modules and the sender.
+             * @param packet[in] Refers to the packet that was sent by the client.
+             * @param client[in] Refers to the client that sent the packet.
+             */
+            virtual void processRequest(std::vector<std::byte> &&packet, IClient const &client) noexcept = 0;
```
Interfaces should not have protected methods

rateau-violent commented on include/IReplySender.hpp:
```diff
@@ -0,0 +1,27 @@
+             * an exception.
+             * @param reply[in] Refers to the data that will be sent to the client.
+             * @param client[in] Refers to the client to send the data to.
+             */virtual void send(ResponseDataType const &reply, IClient const &client) = 0;
```
\n missing

aurelien-boch committed:
feat: Fix required changes in request review

aurelien-boch committed:
feat: add documentation for IPipelineManager

rateau-violent commented on doc/interfaces/IClient.md:
```diff
@@ -0,0 +1,12 @@
+The goal of this interface is to abstract the communication method to the rest of the pipeline.
+
+### Listener
+This interface will be used in the Listener. You are free to chose how your listener works and how it will interact
```
the Listener -> the \`Listener\` ?

Pepiloto replied:
the Listener -> the `Listener`

rateau-violent commented on doc/interfaces/IClient.md:
```diff
@@ -0,0 +1,12 @@
+
+### Listener
+This interface will be used in the Listener. You are free to chose how your listener works and how it will interact
+with the Client.
```
the Client -> the \`Client\` ?

Pepiloto replied:
the Client -> the `Client`

rateau-violent commented on doc/interfaces/IListener.md:
```diff
@@ -0,0 +1,20 @@
+You are free to chose how the internal listener will work (as long as it matches your client implementation obviously).
+
+###  Packet Decryption
+If the packet is encrypted (using SSL / TLS for example) the listener must decrypt it to know when the packet is complete
```
the listener -> the \`Listener\` ?

Pepiloto replied:
listener -> `Listener` ?

rateau-violent commented on doc/interfaces/IListener.md:
```diff
@@ -0,0 +1,20 @@
+and then send the decrypted version to the callback.
+
+### Client lifetime
+The listener is responsible for the lifetime of the `client`. It MUST close it when necessary (for example:
```
listener -> \`Listener\` ?
client > Client ?

Pepiloto replied:
listener -> `listener`

rateau-violent commented on doc/interfaces/IListener.md:
```diff
@@ -0,0 +1,20 @@
+
+### Client lifetime
+The listener is responsible for the lifetime of the `client`. It MUST close it when necessary (for example:
+in HTTP protocol, the `client` must be closed when the reply is sent, whereas in FTP protocol, the `client` must be closed
```
client -> Client

rateau-violent commented on doc/interfaces/IListener.md:
```diff
@@ -0,0 +1,20 @@
+in HTTP protocol, the `client` must be closed when the reply is sent, whereas in FTP protocol, the `client` must be closed
+when the "close" command is received).
+
+*note: in the text above, the `client` term is used, it stands for your IClient implementation which contains a socket.
```
client -> Client

Breigner01 reviewed (approved):
LGTM, just apply the changes asked by Victoire and we are good to go

aurelien-boch committed:
feat: IReplySender.md, IClient.md, IListener.md

Add IReplySender markdown
Fix IListener and IClient, so it matches the pull request requirements

Pepiloto commented on doc/interfaces/IListener.md:
```diff
@@ -0,0 +1,20 @@
+### Client lifetime
+The listener is responsible for the lifetime of the `client`. It MUST close it when necessary (for example:
+in HTTP protocol, the `client` must be closed when the reply is sent, whereas in FTP protocol, the `client` must be closed
+when the "close" command is received).
```
"close" -> `close` ?

aurelien-boch replied:
why should we highlight close ? It's not a client method or a member of our architecture

Breigner01 replied:
I see why he would like to, and I'm not really on one side or the other, I'm ok with both

Pepiloto replied:
Seems legit :+1:

Pepiloto commented on doc/interfaces/IPipelineManager.md:
```diff
@@ -0,0 +1,22 @@
+unused libs. It will also pass an internal method as callback to the `Listener`.
+
+#### The listen state:
+During the listen state, the `PipelineManager` will listen for any new connection and message. When a full message is
```
Message ? not packet ?

Pepiloto commented on doc/interfaces/IReplySender.md:
```diff
@@ -0,0 +1,10 @@
+## Description
+The IReplySender interface defines methods that `ReplySender` must implement.
+The goal of the `ReplySender` is to serialize a reply to its final form according the protocol (for example: in HTTP
+protocol the structure must be converted to a string that meets rfc7540 requirements).
```
rfc7540 -> RFC 7540 ?

Pepiloto commented on doc/interfaces/IReplySender.md:
```diff
@@ -0,0 +1,10 @@
+protocol the structure must be converted to a string that meets rfc7540 requirements).
+
+## How it works
+When the `PipelineManager` is done processing the request and applying all the middlewares, it will call the
```
middlewares or modules ?

Breigner01 commented on doc/interfaces/IClient.md:
```diff
@@ -0,0 +1,16 @@
+with the `Client`.
+
+### ReplySender
+This interface will also be used in the reply sender which will call (normally) only the `send()` method.
```
```suggestion
This interface will also be used in the reply sender which will call (normally) only the `send` method.
```
I'd remove the parenthesis since you state that it's a method

Breigner01 commented on doc/interfaces/IListener.md:
```diff
@@ -0,0 +1,20 @@
+### Description
+The IListener interface defines methods that `Listener` must implement.
+The goal of the `Listener` is to listen for new connections and wait for full packet before calling the callback method,
+that was provided using the `onPacket()` method.
```
```suggestion
that was provided using the `onPacket` method.
```
You state that it is a method so the parenthesis aren't necessary

Breigner01 commented on doc/interfaces/IReplySender.md:
```diff
@@ -0,0 +1,10 @@
+
+## How it works
+When the `PipelineManager` is done processing the request and applying all the middlewares, it will call the
+`send()` method of `IReplySender` with the final reply to send it to client.
```
```suggestion
`send` method of `IReplySender` with the final reply to send it to client.
```
I'd remove the parenthesis since you state that it's a method

Breigner01 reviewed (changes requested):
Few changes for a better readability IMO

Pepiloto commented on readme.md:
```diff
@@ -0,0 +1,94 @@
+
+Pipeline request workflow:
+![Pipeline request workflow](doc/images/request_workflow.png)
+*note that the only difference http and https pipeline is the listener as the pipeline is extremely modular*
```
Maybe explain why is the listener is different

Breigner01 committed:
fix: standardize the puml file

The puml file contained 2 different syntax which were:
- `type name`
- `name: type`
The second one is the one to apply so it's been applied to the whole
file.

aurelien-boch committed:
feat: add virtual destructor to interfaces

aurelien-boch committed:
docs: remove useless parenthesis in IClient.md

Co-authored-by: Benjamin Reigner <benjamin.reigner@epitech.eu>

aurelien-boch committed:
docs: remove useless parenthesis in IListener.md

Co-authored-by: Benjamin Reigner <benjamin.reigner@epitech.eu>

aurelien-boch committed:
docs: remove useless parenthesis in IReplySender.md

Co-authored-by: Benjamin Reigner <benjamin.reigner@epitech.eu>

aurelien-boch committed:
feat: Fix required changes in the pull request review

aurelien-boch committed:
feat: add types documentations, fix CMakeLists.txt

aurelien-boch committed:
docs: remove useless extra line

Breigner01 commented on CMakeLists.txt:
```diff
@@ -24,3 +24,15 @@ target_include_directories(
         include/
 )
+
+if (UNIX)
+    set(COMPILE_FLAGS -Wall -Wextra -Weffc++)
+else()
+    set(COMPILE_FLAGS /W4)
+endif()
```
```suggestion
if (UNIX)
    add_compile_options(-Wall -Wextra -Weffc++)
else ()
    add_compile_options(/W4)
endif ()
```
You can do this instead

aurelien-boch replied:
Using `target_` cmake functions is a good practice as it only edits the targets compile options. `add_compile_options` will add theses options to ALL targets in this directory and below (cf: add_subdirectory)

Pepiloto reviewed (changes requested):
Don't change the name of the contribution file it's for github to show it when people contribute: use CONTRIBUTING.md

aurelien-boch committed:
docs: rename Contributing.md to CONTRIBUTING.md

Pepiloto reviewed (approved):
Good for me
But wait for me for the rebase pls

Breigner01 commented on include/IClient.hpp:
```diff
@@ -23,6 +23,7 @@ namespace api
         #else
             using socket_type = int;
         #endif
+            virtual ~IClient() = default;
```
```suggestion
            virtual ~IClient() = 0;
```
?

Breigner01 commented on include/IListener.hpp:
```diff
@@ -12,6 +12,8 @@ namespace api
     class IListener
     {
         public:
+            virtual ~IListener() = default;
```
```suggestion
            virtual ~IListener() = 0;
```
?

aurelien-boch replied:
There is a difference between = default and = 0:
With = 0 you **MUST** implement the method, with = default, a default value will be assigned by the compiler, here we want a default value.

Breigner01 commented on include/IModule.hpp:
```diff
@@ -15,6 +15,8 @@ namespace api
     class IModule
     {
         public:
+            virtual ~IModule() = default;
```
```suggestion
            virtual ~IModule() = 0;
```
?

aurelien-boch replied:
There is a difference between = default and = 0:
With = 0 you **MUST** implement the method, with = default, a default value will be assigned by the compiler, here we want a default value.

Breigner01 commented on include/IPacketParser.hpp:
```diff
@@ -16,6 +16,8 @@ namespace api
     class IPacketParser
     {
         public:
+            virtual ~IPacketParser() = default;
```
```suggestion
            virtual ~IPacketParser() = 0;
```
?

Breigner01 commented on include/IPipelineManager.hpp:
```diff
@@ -16,6 +16,8 @@ namespace api
     class IPipelineManager
     {
         public:
+            virtual ~IPipelineManager() = default;
```
```suggestion
            virtual ~IPipelineManager() = 0;
```
?

Breigner01 commented on include/IReplySender.hpp:
```diff
@@ -15,12 +15,15 @@ namespace api
     class IReplySender
     {
         public:
+            virtual ~IReplySender() = default;
```
```suggestion
            virtual ~IReplySender() = 0;
```
?

aurelien-boch replied:
There is a difference between = default and = 0:
With = 0 you **MUST** implement the method, with = default, a default value will be assigned by the compiler, here we want a default value.

Breigner01 commented on readme.md:
```diff
@@ -0,0 +1,95 @@
+- Setup phase
+    - The `PipelineManager` contains an instance of the `Listener`, `PacketParser`, `ReplySender` and a vector of all
+      `Modules` used by the pipeline.
+    - The `PipelineManager` will send it's `processRequest()` method as a callback to the `onPacket()` method of the
```
```suggestion
    - The `PipelineManager` will send it's `processRequest` method as a callback to the `onPacket` method of the
```

aurelien-boch replied:
There is a difference between = default and = 0:
With = 0 you **MUST** implement the method, with = default, a default value will be assigned by the compiler, here we want a default value.

Breigner01 commented on readme.md:
```diff
@@ -0,0 +1,95 @@
+      `Modules` used by the pipeline.
+    - The `PipelineManager` will send it's `processRequest()` method as a callback to the `onPacket()` method of the
+      listener
+    - When the `listen()` method of the `Listener` is called, the `Listener` starts listen for new packets.
```
```suggestion
    - When the `listen` method of the `Listener` is called, the `Listener` starts listen for new packets.
```

Breigner01 commented on readme.md:
```diff
@@ -0,0 +1,95 @@
+    - When the `listen()` method of the `Listener` is called, the `Listener` starts listen for new packets.
+
+- Processing phase
+    - When the `Listener` receives a full packet, it just calls corresponding callback (the `processRequest()` method from the
```
```suggestion
    - When the `Listener` receives a full packet, it just calls corresponding callback (the `processRequest` method from the
```

Breigner01 commented on readme.md:
```diff
@@ -0,0 +1,95 @@
+
+Pipeline request workflow:
+![Pipeline request workflow](doc/images/request_workflow.png)
+*note that the only difference http and https pipeline is the listener (the HTTPS one will decrypt SSL)
```
```suggestion
*note that the only difference in http and https pipeline is the listener (the HTTPS one will decrypt SSL)
```

Breigner01 reviewed (changes requested):
Few changes required and we're good to go

aurelien-boch committed:
feat: remove useless parenthesis in readme

Co-authored-by: Benjamin Reigner <benjamin.reigner@epitech.eu>

aurelien-boch committed:
feat: remove useless parenthesis in readme

Co-authored-by: Benjamin Reigner <benjamin.reigner@epitech.eu>

aurelien-boch committed:
docs: update readme.md

aurelien-boch committed:
docs: remove useless parenthesis in readme.md

Breigner01 reviewed (approved):
LGTM
